src/lib/IdLib.sol

```diff
@@ -0,0 +1,378 @@
+        }
+    }
+
+    /**
+     * @notice Internal pure function for converting a reset period to its duration in
+     * seconds. There are eight distinct reset periods ranging from one second to
+     * thirty days. Specific periods include some additional padding:
+     *  - One hour is padded by five minutes
+     *  - Seven days is padded by one hour
+     * @dev No bounds check performed; ensure that the enum value is in range.
+     * @param resetPeriod The reset period to convert.
+     * @return duration   The duration in seconds.
+     */
+    function toSeconds(ResetPeriod resetPeriod) internal pure returns (uint256 duration) {
+        assembly ("memory-safe") {
+            // Bitpacked durations in 24-bit segments:
+            // 278d00  094890  015180  000f3c  000258  00003c  00000f  000001
+            // 30 days 7 days  1 day   1 hour  10 min  1 min   15 sec  1 sec
+            let bitpacked := 0x278d00094890015180000f3c00025800003c00000f000001
+
+            // Shift right by period * 24 bits & mask the least significant 24 bits.
+            duration := and(shr(mul(resetPeriod, 24), bitpacked), 0xffffff)
+        }
+    }
+
+    /**
+     * @notice Internal pure function for computing an address's compact flag. The flag
+     * is a 4-bit value that represents how "compact" the address of an allocator is. A
+     * fully "compact" allocator address will have nine leading zero bytes, or eighteen
+     * leading zero nibbles. To be considered even partially compact, the account must
+     * have at least two leading zero bytes, or four leading zero nibbles. The full
+     * scoring formula is therefore:
+     *  - 0-3 leading zero nibbles: 0
+     *  - 4-17 leading zero nibbles: number of leading zeros minus 3
+     *  - 18+ leading zero nibbles: 15
+     * @param allocator    The address to compute the flag for.
+     * @return compactFlag The computed compact flag.
+     */
+    function toCompactFlag(address allocator) internal pure returns (uint8 compactFlag) {
+        assembly ("memory-safe") {
+            // Extract the uppermost 72 bits of the address.
+            let x := shr(168, shl(96, allocator))
+
+            // Propagate the highest set bit.
+            x := or(x, shr(1, x))
+            x := or(x, shr(2, x))
+            x := or(x, shr(4, x))
+            x := or(x, shr(8, x))
+            x := or(x, shr(16, x))
+            x := or(x, shr(32, x))
+
+            // Count set bits to derive most significant bit in the last byte.
+            let y := sub(x, and(shr(1, x), 0x5555555555555555))
+            y := add(and(y, 0x3333333333333333), and(shr(2, y), 0x3333333333333333))
+            y := and(add(y, shr(4, y)), 0x0f0f0f0f0f0f0f0f)
+            y := add(y, shr(8, y))
+            y := add(y, shr(16, y))
+            y := add(y, shr(32, y))
+
+            // Look up final value in the sequence.
+            compactFlag := and(shr(and(sub(72, and(y, 127)), not(3)), 0xfedcba9876543210000), 15)
```

Comment:
Can we add some sort of mathematical proof to show, that the last 88 bits of an address + the `compactFlag` is sufficient to be fully unique?

Reply:
would love that (assuming you mean as a comment or a test, since this is getting evaluated frequently)

Reply:
Yes, that is what I mean